### ISSUE TITLE
Don't read configuration if not interactive

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#5394]
One of the nice things of fish is that we don't have anything as complicated as the bash man page's "INVOCATION" section in our documentation. 

"If it's a non-interactive shell, it does not read bashrc, except when it has been invoked by ssh or rshd, unless it is invoked as sh".

We always read config.fish, and that's nice.

However, as we see in e.g. #5246 and #4432, doing that even when non-interactive can cause some issues. Issues that we currently expect the user to expect, and guard against with `if status is-interactive`.

And I'm not really sure if configuration of non-interactive shells is ever really wanted? If you execute a script via `fish /path/to/script` (or directly via the shebang), you expect a consistent environment.

So, I would argue that we should possibly change this:

- Don't source /etc/fish/config.fish and ~/.config/fish/config.fish (the one in /usr can stay, but that's not really "config")

- Don't read snippets (which are technically in /usr/share/fish/config.fish)

- Don't add the user's functions (in $HOME and /etc) to $fish_function_path

Now, obviously this complicates the logic a bit. It's "if fish is interactive, read config.fish, if not, don't" instead of just "read config.fish". I'd argue that it's still quite simple, and helps unbreak things that are currently quite easy to break - understanding that `stty` touches the terminal is advanced stuff.

Additonally, if configuration of non-interactive shells is wanted (and I don't see how, given that bash doesn't have it), we could add a "config-noninteractive.fish", or simply expect users to run `fish -C 'source ~/.config/fish/config.fish`.

**Comment by zanchey** [comment]
I'm against this. I do have stuff in my non-interactive startup config (path editing dependent on machine, for example), and I use my custom functions from non-interactive mode all the time.

**Comment by faho** [comment]
>I do have stuff in my non-interactive startup config (path editing dependent on machine, for example)

Could you provide some more detail?

Like I said, it's _possible_ that non-interactive configuration is needed, I just believe that the default is backwards. Most configuration should not apply to non-interactive shells, and the current solution is a footgun.

>path editing dependent on machine

You mean $PATH? Shouldn't that be done once in the login shell and then passed down? Or in e.g. pam_env? (There's a bit of an annoyance here in that much of linux expects a shell to run in the startup path, and specifically one that can read /etc/profile* - but fish can't do that to begin with) 

>I use my custom functions from non-interactive mode all the time.

That's the one I was least sure about. Do users expect functions to be available? They aren't as much of an issue as the configuration is (autoloading prevents them from doing stuff to the terminal in the startup path, unless that calls them), so it might also be okay to leave $fish_function_path unchanged.

Though that of course means it's different between interactive and non-interactive shells if your configuration changes $fish_function_path (mine does, but explicitly for interactive shells).

**Comment by ridiculousfish** [comment]
I'm against this too for the reason given in the intro. If we don't source config.fish, then we will get requests for how to configure fish in non-interactive mode, and there will be pressure to add a non-interactive config file, and then we're well on the road to bash.

It's true that surprising consequences erupt when configs designed for interactive mode are sourced in non-interactive mode. I believe it's worth tolerating this to maintain the simple initialization story.

Now there exists a function `__fish_config_interactive`, and it seems OK for it to invoke a function `fish_user_config_interactive` or similar. We can then document that interactive-only configs should go in this function. That seems better to me: when initialization happens in fish script then it is easier for users to inspect and understand.

**Comment by faho** [comment]
>it seems OK for it to invoke a function fish_user_config_interactive or similar.

That doesn't seem much better than `if status is-interactive`.

Plus, the problem I see is really one of defaults. Currently, everything in the config is executed in non-interactive fish _by default_. You need to do things to stop that from happening. So you need to do things in the majority of cases, even if you don't even know that you need to do things.

I'd like to figure out a way to turn that around.

**Comment by mqudsi** [comment]
See also #5417 which is at least gives the user the option of explicitly opting out of loading `config.fish`.

**Comment by faho** [comment]
So I'd like to get back to this.

That's because this is necessary if we expect fish to be used via shebanged scripts.

And in that case, a commandline option to fish doesn't work, because fish might be opened via /usr/bin/env, and linux doesn't split any arguments beyond the first - so `#!/usr/bin/env fish --no-config` would try to open a file called `fish --no-config`.

If never reading config.fish for non-interactive shells is unacceptable, there are other ideas we could talk about, including:

- Use config-noninteractive.fish if that exists instead of config.fish, so users can `touch` it to remove their customization from non-interactive mode
- Allow config-interactive.fish so users can use that
- Use config-noninteractive.fish in non-interactive mode, so users have to add `source config.fish` to that to enable the same customizations in non-interactive mode

(these would also include setting $fish_function_path and reading snippets - so we'd have to add a bit to read the snippets to the third option)

**Comment by zanchey** [comment]
> > I do have stuff in my non-interactive startup config (path editing dependent on machine, for example)
> 
> Could you provide some more detail?

For example:
 * [keychain](https://www.funtoo.org/Keychain)
 * path management, which I can't set up using the login mechanism (no root) or `pam_env` (not available)

**Comment by krobelus** [comment]
I think it's a great default that `fish -c foo` behaves just like interactive `foo`.
Sane `config.fish`/functions/snippets should not break reasonable scripts, but I guess script authors don't want to rely on that..

> And in that case, a commandline option to fish doesn't work, because fish might be opened via /usr/bin/env, and linux doesn't split any arguments beyond the first - so `#!/usr/bin/env fish --no-config` would try to open a file called `fish --no-config`.

The `-S` (split-string) option to `env` should be available, right?

**Comment by faho** [comment]
>The -S (split-string) option to env should be available, right?

Not in Posix and... basically nobody uses it, or thinks of using it? Even if people figured that `fish --norc` should be used (they most likely won't), the chance of somebody finding `env -S` is not good.

So for portable scripts it's out (discounting of course that /usr/bin/env is technically not even guaranteed to be the path to env, but I have zero sympathy for systems that don't have that).

> Sane config.fish/functions/snippets should not break reasonable scripts, but I guess script authors don't want to rely on that..

Most people *don't* think of non-interactive scripts when configuring their shell for interactive use. We've had people make a `test` function and not notice it for a bit. Making e.g. a `grep` alias that does `--color=always` would be even harder to find.

Also plenty of configurations just do some slow interactive setup, and that's just wasteful and makes fish for scripts appear worse.

Basically it's needless footguns, is my position.

----

(Also, weirdly enough keychain states "Currently, keychain is not compatible with Fish shell.")

**Comment by InspectorMustache** [comment]
I just stumbled across this and have to say, I would be in hard favour of not sourcing config.fish whenever fish is launched non-interactively. The way it is now, it is very hard to write scripts that can be shared in any way because you'd always have to accomodate for potential quirks in some user's config. This effectively makes fish script useless as a general scripting language for anything other than customising your *own* shell.
If you really need one or two of your functions in your personal scripts, couldn't you just source those functions from within the script and then run them? Not very elegant, I admit, but that script is designed to be used by noone but yourself anyway.

**Comment by Korkman** [comment]
How about a custom she-bang
```
#! /usr/bin/fish_script
```
typically symlinked to fish and if it is called like that, skip config?

I think the current default loading the system config is fine. And I would always assume someone willing to write portable scripts to go an extra mile in the documentation to find such a quirk. Ensuring portability is hard anyways because shell scripts rely so heavily on external programs and their expected behavior.

**Comment by IlanCosman** [comment]
I recently ran into this issue. I'm amazed I didn't think of `status is-interactive` until reading this thread, I was doing all sorts of contortions to avoid loading stuff. 

I think that non-interactive `fish` should still load the config by default, but that there should be a `fish` flag to not do so.

**Comment by InspectorMustache** [comment]
An unpredictable interpreter just seems like bad practice to me.  Sure, if someone really wants to write portable scripts in Fish, they can sprinkle `command`s everywhere and put a notice in the documentation that their script won't work if your fish config contains "something weird".  Why would they want to though?  In that case, they might as well just use bash.

For me, this isn't about whether or not someone who's dream it is to write portable Fish scripts can or cannot do so.  It's about what the sensible design choice is.  I guess, another question is whether fish script is intended to be used for shell scripting or just as a DSL to customise your shell.  For shell scripting, I feel a predictable environment is clearly preferable.  You can always use functions if you want to rely on your config.  Of course, if it's just intended for shell customisation, it doesn't really matter.  I like to use fish script for general shell scripting occasionally, but of course that's just a personal preference.

**Comment by realmazharhussain** [comment]
**Simple solution:**
Do it like bash;
0. Add a /etc/profile.fish and ~/.fish_profile or ~/.config/profile.fish.
1. These files should run at login time only
2. Only these files should run when non-interactive login
3. All files (config files + profile.fish files) should run when interactive login
4. Nothing should run when non-login and non-interactive

**In simple words:**
when and only when `status is-login`, fish runs profile.fish or .fish_profile
when and only when `status is-interactive`, fish runs config files and all other stuff

**Advantages:**
- Fish is consistent across almost all fish configs and setups
- Important fish customizations can be done once at login time like setting system variables e.g. PATH, XDG_DATA_DIRS, etc.

**Comment by faho** [comment]
>Add /etc/config-interactive.fish and corresponding user-specific file

@realmazharhussain Please read back (https://github.com/fish-shell/fish-shell/issues/5394#issuecomment-678393660), that was my proposal 5. 

My criticism stands. Most configuration is either useless or harmful in a non-interactive context, so the default should be to not read configuration when non-interactive. 

Giving people another thing they can use if they think of using it doesn't help - we already have those!

**Comment by realmazharhussain** [comment]
@faho I have read your comment and I still think my purposal is the right way to go.

Coreutils 8.30 and newer have env that supports -S option so you can write `#!/usr/bin/env -S fish --no-config` as shebang and it will work perfectly when new version of fish comes out.

If you want to ask "But what about old distros like Debian stable, etc? They don't have newer version of coreutils. My script is not going to work on those distros.", You know what, they are not gonna get new and fixed version of fish either. So, your script won't work on those distros either way.
Then you might say "People on those distros can just install fish from source.". I would say most the people won't and your script would still not work if you distribute your script to the public.

If you write a script for yourself, you know exactly where your fish executable is. Just use that as shebang instead of /usr/bin/env.
Your `#!/path/to/fish --no-config` will work perfectly.

**Comment by faho** [comment]
>Coreutils 8.30 and newer have env that supports -S option

@realmazharhussain  We do not only support GNU coreutils. We do not only support linux, we do not only support new linux distributions.

We don't want users to have to remember to use the fairly obscure `-S` switch.

This has been discussed before, you aren't adding anything new, please stop.

**Comment by realmazharhussain** [comment]
> This has been discussed before, you aren't adding anything new, please stop.

@faho 
I actually did add something; the idea of adding config-interactive.fish instead of config-noninteractive.fish.
It's OK if you don't like my suggestion. But since you prefer to talk like this, I'll stop now.

I write my scripts in bash anyway. AND I have no issue with fish the way it currently works 'cause I only use it interactively.

**Comment by madprops** [comment]
I'm not sure if related. But I'm trying to figure out why the Dolphin's (file manager) integrated terminal (which I think is Konsole underneath) doesn't seem to load a full fish shell. For instance my aliases don't work. If I do `exit` and start the terminal again, they do work. Also sometimes (not sure how to trigger it) I see errors like `fish: Unknown command: fish_git_prompt`

**Comment by faho** [comment]
@madprops No, that's entirely unrelated.

**Comment by philn128** [comment]
Just ran into this today. I had an environment variable in my `~/.config/fish/config.fish` that was overriding an environment variable I had set; definitely unexpected behavior. IF someone actually wanted to run `config.fish` all they'd have to do is

```bash
#!/usr/bin/env fish
source $__fish_config/config.fish
```

Nonetheless, I now have at the top of my `config.fish`

```bash
if not status --is-interactive
	exit
end
```

**Comment by lgarron** [comment]
I ran into this issue in a GitHub codespace with `fish` 3.6.1.

`#!/usr/bin/env fish` does not load `config.fish` on macOS, and I don't recall it doing so in codespaces in the past. But it's doing so now!

> > And in that case, a commandline option to fish doesn't work, because fish might be opened via /usr/bin/env, and linux doesn't split any arguments beyond the first - so `#!/usr/bin/env fish --no-config` would try to open a file called `fish --no-config`.
> 
> The `-S` (split-string) option to `env` should be available, right?

This workaround seems to work for me for now.

I've written shell scripts  for a long time now and never knew about that caveat, so this sure seems extremely non-obvious.

But these scripts are not supposed to load `config.fish`, so the `status --is-interactive` workaround would be insufficient if anyone else ever used them.

I guess I'm fine with this, but is there a good explanation for why `/usr/bin/env fish` doesn't seem to load `config.fish` on macOS? And why I would not expect this to be consistent across POSIX-like systems?

**Comment by lgarron** [comment]
> I've written shell scripts for a long time now and never knew about that caveat, so this sure seems extremely non-obvious.

In fact, seems like this is *so* non-obvious that none of the `fish` shell scripts in this repo do it:

https://github.com/search?q=repo%3Afish-shell%2Ffish-shell+%2F%23%21%5C%2Fusr%5C%2Fbin%5C%2Fenv.*fish%2F&type=code

**Comment by faho** [comment]
>In fact, seems like this is so non-obvious that none of the fish shell scripts in this repo do it:

They can't do it because `env` might not have a `-S` option.

It's not just that it's non-obvious, it's that there is no portable way to do it.

>I guess I'm fine with this, but is there a good explanation for why /usr/bin/env fish doesn't seem to load config.fish on macOS?

It should, I don't know why it wouldn't. Maybe $XDG_CONFIG_HOME is set to something in that context?

**Comment by lgarron** [comment]
> They can't do it because `env` might not have a `-S` option.
> 
> It's not just that it's non-obvious, it's that there is no portable way to do it.

Ah, I see. It would definitely be very useful to have a portable option.

> It should, I don't know why it wouldn't. Maybe $XDG_CONFIG_HOME is set to something in that context?

Looks like it's empty. 🤷‍♀️

**Comment by faho** [comment]
>Looks like it's empty. 🤷‍♀️

Try setting $FISH_DEBUG to "config" and then run the file. E.g.

```fish
FISH_DEBUG=config ./foo.fish
```

This will print output like:

>config: exec_path: "/usr/local/bin/fish", argv[0]: "fish"
config: Using compiled in paths:
config: determine_config_directory_paths() results:
paths.data: /usr/local/share/fish
paths.sysconf: /usr/local/etc/fish
paths.doc: /usr/local/share/doc/fish
paths.bin: /usr/local/bin
config: sourcing /usr/local/share/fish/config.fish
config: sourcing /usr/local/etc/fish/config.fish
config: sourcing /home/user/.config/fish/config.fish

Those last three lines will tell you exactly which files it ran. If it did, check the files to see if you have a `status is-interactive` in there somewhere.

**Comment by lgarron** [comment]
Aha, thanks!

Looks like it was always (unintentionally) loading `config.fish` on macOS as well as codespaces. It's not until recently that something in Codespaces changed so that `config.fish` would print something for me on first run! 😳

```
> FISH_DEBUG=config git-distance origin/main
config: exec_path: '/opt/homebrew/Cellar/fish/3.6.1/bin/fish', argv[0]: 'fish'
config: Using compiled in paths:
config: determine_config_directory_paths() results:
paths.data: /opt/homebrew/Cellar/fish/3.6.1/share/fish
paths.sysconf: /opt/homebrew/etc/fish
paths.doc: /opt/homebrew/Cellar/fish/3.6.1/share/doc/fish
paths.bin: /opt/homebrew/Cellar/fish/3.6.1/bin
config: sourcing /opt/homebrew/Cellar/fish/3.6.1/share/fish/config.fish
config: sourcing /opt/homebrew/etc/fish/config.fish
config: sourcing /Users/lgarron/.config/fish/config.fish
WARNING: Loaded `config.fish` outside an interactive shell.
0 commit(s) on main
0 commit(s) on origin/main
```
